File: 1209-remove-all-adjacent-duplicates-in-string-ii/1209-remove-all-adjacent-duplicates-in-string-ii.cpp
class Solution {
public:
    string removeDuplicates(string s, int k) {
        int n=s.size();
        stack<pair<char,int>> stk;
        for(int i=0;i<n;i++){
            if(stk.empty() or stk.top().first!=s[i]) stk.push({s[i],1});
            else{
                auto prev=stk.top();
                stk.pop();
                stk.push({s[i],prev.second+1});
            }
            if(stk.top().second==k){
                stk.pop();
            }
        }
        string result="";
        while(!stk.empty()){
            auto temp=stk.top();
            while(temp.second-- > 0){
                result.push_back(temp.first);
            }
            stk.pop();
        }
        reverse(result.begin(),result.end());
        return result;
    }
};

File: 1209-remove-all-adjacent-duplicates-in-string-ii/1209-remove-all-adjacent-duplicates-in-string-ii.java
class Solution {
    public String removeDuplicates(String s, int k) {
        Stack<Pair<Character,Integer>> stack=new Stack<>();
        for(int i=0; i<s.length(); i++){
            char ch=s.charAt(i);
            if(stack.isEmpty() || stack.peek().getKey()!=ch){
                stack.push(new Pair<>(ch, 1));
            }else{
                if(stack.peek().getValue()==k-1){
                    stack.pop();
                }else{
                    Pair<Character,Integer> top=stack.pop();
                    stack.push(new Pair<>(top.getKey(), top.getValue()+1));
                }
            }
        }
        StringBuilder sb=new StringBuilder();
        while(!stack.isEmpty()){
            Pair<Character,Integer> top=stack.pop();
            for(int i=0; i<top.getValue(); i++){
                sb.insert(0, top.getKey());
            }
        }
        return sb.toString();
    }
}